lib/spack/spack/spec.py

```diff
@@ -2933,15 +2924,21 @@ def __getitem__(self, name):
         return value
 
     def __contains__(self, spec):
-        """True if this spec satisfies the provided spec, or if any dependency
-           does.  If the spec has no name, then we parse this one first.
+        """True if this spec or some dependency satisfies the spec.
+
+        Note: If ``spec`` is anonymous, we ONLY check whether the root
+        satisfies it, NOT dependencies.  This is because most anonymous
+        specs (e.g., ``@1.2``) don't make sense when applied across an
+        entire DAG -- we limit them to the root.
+
         """
         spec = self._autospec(spec)
-        for s in self.traverse():
-            if s.satisfies(spec, strict=True):
-                return True
 
-        return False
+        # if anonymous or same name, we only have to look at the root
+        if not spec.name or spec.name == self.name:
+            return self.satisfies(spec)
+        else:
+            return any(s.satisfies(spec) for s in self.traverse(root=False))
```

Comment:
@tgamblin Do you remember if there was any reason for passing from `strict=True` here to `strict=False`?